contracts/CompoundFCM.sol

```diff
@@ -0,0 +1,337 @@
+
+    require(notional!=0, "notional = 0");
+
+    /// todo: (URGENT) add support for approvals and recipient (similar to how it is implemented in the MarginEngine)
+
+    // initiate a swap
+    // the default tick range for a Position associated with the FCM is tickLower: -tickSpacing and tickUpper: tickSpacing
+    // isExternal is true since the state updates following a VAMM induced swap are done in the FCM (below)
+    IVAMM.SwapParams memory params = IVAMM.SwapParams({
+        recipient: address(this),
+        amountSpecified: notional.toInt256(),
+        sqrtPriceLimitX96: sqrtPriceLimitX96,
+        tickLower: -tickSpacing,
+        tickUpper: tickSpacing
+    });
+
+    (int256 fixedTokenDelta, int256 variableTokenDelta, uint256 cumulativeFeeIncurred, ,) = _vamm.swap(params);
+
+    require(variableTokenDelta <=0, "VT delta sign");
+
+    TraderWithYieldBearingAssets.Info storage trader = traders[msg.sender];
+
+    uint256 currentExchangeRate = _ctoken.exchangeRateCurrent();
+
+    /// @audit-casting variableTokenDelta is expected to be negative here, but what if goes above 0 due to rounding imprecision?
+    uint256 yieldBearingTokenDelta = uint256(-variableTokenDelta).wadDiv(currentExchangeRate);
+    uint256 updatedTraderMargin = trader.marginInScaledYieldBearingTokens + yieldBearingTokenDelta;
+    trader.updateMarginInScaledYieldBearingTokens(updatedTraderMargin);
+
+    // update trader fixed and variable token balances
+    trader.updateBalancesViaDeltas(fixedTokenDelta, variableTokenDelta);
+
+    // deposit notional executed in terms of cTokens (e.g. cDAI) to fully collateralise the position
+    // we need a number of tokens equal to the variable token delta divided by the exchange rate
+    IERC20Minimal(address(_ctoken)).safeTransferFrom(msg.sender, address(this), yieldBearingTokenDelta);
+
+    // transfer fees to the margin engine (in terms of the underlyingToken e.g. cDAI)
+    underlyingToken.safeTransferFrom(msg.sender, address(_marginEngine), cumulativeFeeIncurred);
+
+    emit FullyCollateralisedSwap(msg.sender, trader.marginInScaledYieldBearingTokens, trader.fixedTokenBalance, trader.variableTokenBalance);
+  }
+
+  /// @notice Get Trader Margin In Yield Bearing Tokens
+  /// @dev this function takes the scaledBalance associated with a trader and multiplies it by the current Exchange Rate to get the balance (margin) in terms of the underlying token
+  /// @param traderMarginInScaledYieldBearingTokens traderMarginInScaledYieldBearingTokens
+  function getTraderMarginInYieldBearingTokens(uint256 traderMarginInScaledYieldBearingTokens) internal view returns (uint256 marginInYieldBearingTokens) {
+    // uint256 currentExchangeRate = _ctoken.exchangeRateStored();
+    // marginInYieldBearingTokens = traderMarginInScaledYieldBearingTokens.rayMul(currentExchangeRate);
+    return traderMarginInScaledYieldBearingTokens; // NO scaling to do here. Delete this function?
```

Comment:
yeah, this feels like a redundant function

Reply:
why is there no scaling to do here? i think we need to scale to convert cTokens into Tokens?

Reply:
this function should be the traderMarginInUnderlyingTokens! In case of aave it works since 1 aToken = 1 Token and getTraderMarginInYieldBearingTokens is equivalent to getTraderMarginInUnderlyingTokens